presto-product-tests/src/main/java/io/prestosql/tests/kafka/TestKafkaPushdownSmokeTest.java

```diff
@@ -0,0 +1,218 @@
+                KAFKA_CATALOG,
+                SCHEMA_NAME,
+                PUSHDOWN_OFFSET_TABLE_NAME)))
+                .containsExactly(row(12));
+
+        // TODO Assert from the query stats that only 10 messages were read
+        assertThat(query(format(
+                "SELECT COUNT(*) FROM %s.%s.%s WHERE _partition_offset >= 5 AND _partition_offset < 10",
+                KAFKA_CATALOG,
+                SCHEMA_NAME,
+                PUSHDOWN_OFFSET_TABLE_NAME)))
+                .containsExactly(row(10));
+
+        // TODO Assert from the query stats that only 10 messages were read
+        assertThat(query(format(
+                "SELECT COUNT(*) FROM %s.%s.%s WHERE _partition_offset > 5 AND _partition_offset <= 10",
+                KAFKA_CATALOG,
+                SCHEMA_NAME,
+                PUSHDOWN_OFFSET_TABLE_NAME)))
+                .containsExactly(row(10));
+
+        // TODO Assert from the query stats that only 2 messages were read
+        assertThat(query(format(
+                "SELECT COUNT(*) FROM %s.%s.%s WHERE _partition_offset = 5",
+                KAFKA_CATALOG,
+                SCHEMA_NAME,
+                PUSHDOWN_OFFSET_TABLE_NAME)))
+                .containsExactly(row(2));
+    }
+
+    private static class PushdownCreateTimeTable
+            implements RequirementsProvider
+    {
+        @Override
+        public Requirement getRequirements(Configuration configuration)
+        {
+            return immutableTable(new KafkaTableDefinition(
+                    SCHEMA_NAME + "." + PUSHDOWN_CREATE_TIME_TABLE_NAME,
+                    PUSHDOWN_CREATE_TIME_TOPIC_NAME,
+                    new ListKafkaDataSource(ImmutableList.of()),
+                    1,
+                    1));
+        }
+    }
+
+    @Test(groups = {KAFKA, PROFILE_SPECIFIC_TESTS})
+    @Requires(PushdownCreateTimeTable.class)
+    public void testCreateTimePushdown()
+            throws InterruptedException
+    {
+        // Ensure a spread of at-least TIMESTAMP_NUM_MESSAGES * 100 milliseconds
+        for (int i = 1; i <= TIMESTAMP_NUM_MESSAGES; i++) {
+            query(format("INSERT INTO %s.%s.%s (bigint_key, bigint_value) VALUES (%s, %s)",
+                    KAFKA_CATALOG, SCHEMA_NAME, PUSHDOWN_CREATE_TIME_TABLE_NAME, i, i));
+            Thread.sleep(100);
+        }
+
+        long startKey = 4;
+        long endKey = 6;
+        List<List<?>> rows = query(format(
+                "SELECT CAST(_timestamp AS VARCHAR) FROM %s.%s.%s WHERE bigint_key IN (" + startKey + ", " + endKey + ") ORDER BY bigint_key",
+                KAFKA_CATALOG,
+                SCHEMA_NAME,
+                PUSHDOWN_CREATE_TIME_TABLE_NAME))
+                .rows();
+        String startTime = (String) rows.get(0).get(0);
+        String endTime = (String) rows.get(1).get(0);
+
+        // TODO Assert from the query stats that only TIMESTAMP_NUM_MESSAGES - startKey messages were read because upper bound is not pushed down
```

Comment:
I think we decided to skip that. If we want a followup please create an issue for that and add as //todo parameter. 
Drop whole comment otherwise.

Reply:
Will drop this comment. I too don't see any possible way to fulfill this TODO.